src/ompl/geometric/planners/experience/src/ThunderRetrieveRepair.cpp

```diff
@@ -0,0 +1,456 @@
+
+#include <boost/thread.hpp>
+
+#include <limits>
+
+ompl::geometric::ThunderRetrieveRepair::ThunderRetrieveRepair(const base::SpaceInformationPtr &si, const ompl::tools::ThunderDBPtr &experienceDB)
+    : base::Planner(si, "Thunder_Retrieve_Repair"),
+      experienceDB_(experienceDB),
+      nearestK_(10) // default value
+{
+    specs_.approximateSolutions = true;
+    specs_.directed = true;
+
+    // Repair Planner Specific:
+    repairProblemDef_.reset(new base::ProblemDefinition(si_));
+
+    psk_.reset(new ompl::geometric::PathSimplifier(si_));
+}
+
+ompl::geometric::ThunderRetrieveRepair::~ThunderRetrieveRepair(void)
+{
+    freeMemory();
+}
+
+void ompl::geometric::ThunderRetrieveRepair::clear(void)
+{
+    Planner::clear();
+    freeMemory();
+
+    // Clear the inner planner
+    if (repairPlanner_)
+        repairPlanner_->clear();
+}
+
+void ompl::geometric::ThunderRetrieveRepair::setExperienceDB(ompl::tools::ThunderDBPtr experienceDB)
+{
+    experienceDB_ = experienceDB;
+}
+
+void ompl::geometric::ThunderRetrieveRepair::setRepairPlanner(const base::PlannerPtr &planner)
+{
+    if (planner && planner->getSpaceInformation().get() != si_.get())
+        throw Exception("Repair planner instance does not match space information");
+    repairPlanner_ = planner;
+    setup_ = false;
+}
+
+void ompl::geometric::ThunderRetrieveRepair::setup(void)
+{
+    Planner::setup();
+
+    // Setup repair planner (for use by the rrPlanner)
+    // Note: does not use the same pdef as the main planner in this class
+    if (!repairPlanner_)
+    {
+        // Set the repair planner
+        boost::shared_ptr<ompl::geometric::RRTConnect> repair_planner( new ompl::geometric::RRTConnect( si_ ) );
+        //repair_planner->setGoalBias(0.2);
+
+        OMPL_DEBUG("No repairing planner specified. Using default: %s", repair_planner->getName().c_str() );
+        repairPlanner_ = repair_planner; //ompl::geometric::Planner( repair_planer );
+    }
+    // Setup the problem definition for the repair planner
+    repairProblemDef_->setOptimizationObjective(pdef_->getOptimizationObjective()); // copy primary problem def
+
+    // Setup repair planner
+    repairPlanner_->setProblemDefinition(repairProblemDef_);
+    if (!repairPlanner_->isSetup())
+        repairPlanner_->setup();
+
+
+}
+
+void ompl::geometric::ThunderRetrieveRepair::freeMemory(void)
+{
+}
+
+ompl::base::PlannerStatus ompl::geometric::ThunderRetrieveRepair::solve(const base::PlannerTerminationCondition &ptc)
+{
+    bool solved = false;
+    double approxdif = std::numeric_limits<double>::infinity();
+    nearestPaths_.clear();
+
+    // Check if the database is empty
+    if (experienceDB_->isEmpty())
+    {
+        std::cout << ANSI_COLOR_GREEN;
+        OMPL_INFORM("Experience database is empty so unable to run ThunderRetrieveRepair algorithm.");      
+        std::cout << ANSI_COLOR_RESET;
+
+        // TODO: it seems TIMEOUT causes the wrong behavior in parallel plan
```

Comment:
I have no idea anymore. I changed comment to

> // Using CRASH because it seems TIMEOUT causes the wrong behavior in parallel plan, not sure why anymore